File: backend/src/main/java/com/popcorn/demo/domain/payment/controller/PaymentController.java
package com.popcorn.demo.domain.payment.controller;

import java.util.List;
import java.util.UUID;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.popcorn.demo.common.controller.BaseController;
import com.popcorn.demo.common.dto.BaseResponse;
import com.popcorn.demo.common.versioning.ApiVersion;
import com.popcorn.demo.domain.payment.dto.request.PaymentStatusUpdateRequest;
import com.popcorn.demo.domain.payment.dto.response.PaymentDetailResponse;
import com.popcorn.demo.domain.payment.dto.response.PaymentListResponse;
import com.popcorn.demo.domain.payment.service.PaymentCommandService;
import com.popcorn.demo.domain.payment.service.PaymentQueryService;
import com.popcorn.demo.domain.payment.service.PaymentQueryService.PaymentDetailResult;
import com.popcorn.demo.common.annotation.ApiLogging;
import com.popcorn.demo.common.annotation.AuditLog;
import com.popcorn.demo.common.annotation.CacheResult;
import com.popcorn.demo.common.annotation.Idempotent;
import com.popcorn.demo.common.annotation.RateLimit;
import com.popcorn.demo.common.annotation.RetryOnFailure;
import com.popcorn.demo.common.annotation.ValidateRequest;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.ExampleObject;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;

@RestController
@Tag(name = "Payments", description = "결제 관리 API")
@ApiVersion("v1")
@RequestMapping("/api/v1")
@RequiredArgsConstructor
public class PaymentController extends BaseController {

	private final PaymentCommandService paymentCommandService;
	private final PaymentQueryService paymentQueryService;

	@Operation(summary = "결제 조회(주문 기준)", description = "주문 기준으로 결제 기록을 조회합니다.")
	@ApiResponse(
			responseCode = "200",
			description = "결제 목록 조회 성공",
			content = @Content(schema = @Schema(implementation = PaymentListResponse.class))
	)
	@GetMapping("/orders/{orderId}/payments")
	@CacheResult(
		cacheName = "paymentsByOrderCache",
		keyExpression = "#orderId",
		ttlSeconds = 180,
		condition = "#orderId != null"
	)
	@ApiLogging(
		message = "주문별 결제 목록 조회",
		includeRequest = true,
		includeResponse = false,
		level = ApiLogging.LogLevel.INFO
	)
	@RateLimit(
		requests = 20,
		window = 60,
		keyExpression = "T(org.springframework.web.context.request.RequestContextHolder).currentRequestAttributes().getRequest().getRemoteAddr()",
		errorMessage = "결제 조회 요청이 너무 많습니다."
	)
	@RetryOnFailure(
		maxAttempts = 2,
		backoffMillis = 300,
		retryOn = {RuntimeException.class},
		logRetryAttempts = false
	)
	public ResponseEntity<BaseResponse<PaymentListResponse>> getPaymentsByOrder(
			@Parameter(description = "주문 ID", required = true, example = "40000000-0000-0000-0000-000000000004")
			@PathVariable UUID orderId) {
		List<PaymentDetailResult> results = paymentQueryService.getPaymentsByOrder(orderId);
		List<PaymentListResponse.Item> items = results.stream()
				.map(this::toListItem)
				.toList();
		PaymentListResponse response = PaymentListResponse.builder()
				.count(items.size())
				.items(items)
				.build();
		return ok(response);
	}

	@Operation(summary = "결제 단건 조회", description = "결제 ID로 결제 기록을 조회합니다.")
	@ApiResponse(
			responseCode = "200",
			description = "결제 단건 조회 성공",
			content = @Content(schema = @Schema(implementation = PaymentDetailResponse.class))
	)
	@GetMapping("/payments/{paymentId}")
	@CacheResult(
		cacheName = "paymentDetailCache",
		keyExpression = "#paymentId",
		ttlSeconds = 120,
		condition = "#paymentId != null"
	)
	@ApiLogging(
		message = "결제 상세 조회",
		includeRequest = true,
		includeResponse = false,
		level = ApiLogging.LogLevel.INFO
	)
	@RateLimit(
		requests = 30,
		window = 60,
		keyExpression = "T(org.springframework.web.context.request.RequestContextHolder).currentRequestAttributes().getRequest().getRemoteAddr()",
		errorMessage = "결제 상세 조회 요청이 너무 많습니다."
	)
	@RetryOnFailure(
		maxAttempts = 2,
		backoffMillis = 300,
		retryOn = {RuntimeException.class},
		logRetryAttempts = false
	)
	public ResponseEntity<BaseResponse<PaymentDetailResponse>> getPayment(
			@Parameter(description = "결제 ID", required = true, example = "70000000-0000-0000-0000-000000000001")
			@PathVariable UUID paymentId) {
		PaymentDetailResult result = paymentQueryService.getPayment(paymentId);
		return ok(toDetailResponse(result));
	}

	@Operation(
			summary = "결제 상태 변경",
			description = """
				결제 상태를 변경합니다.

				**상태 전이 규칙:**
				- READY → PAID, FAILED, CANCELLED
				- PAID → CANCELLED
				- FAILED → (최종 상태)
				- CANCELLED → (최종 상태)

				**상태 변경 시 동작:**
				- PAID: 승인 시각 저장 및 주문 상태 갱신
				- CANCELLED: 주문 상태를 CANCELLED로 변경
				"""
	)
	@ApiResponse(
			responseCode = "200",
			description = "결제 상태 변경 성공",
			content = @Content(schema = @Schema(implementation = PaymentDetailResponse.class))
	)
	@PatchMapping("/payments/{paymentId}/status")
	@RateLimit(
		requests = 5,
		window = 60,
		keyExpression = "T(org.springframework.web.context.request.RequestContextHolder).currentRequestAttributes().getRequest().getRemoteAddr()",
		errorMessage = "🚨 결제 상태 변경은 분당 5회만 허용됩니다.",
		algorithm = RateLimit.Algorithm.SLIDING_WINDOW
	)
	@ApiLogging(
		message = "🔒 결제 상태 변경",
		includeRequest = true,
		includeResponse = true,
		includeExecutionTime = true,
		maskSensitiveData = true,
		level = ApiLogging.LogLevel.WARN
	)
	@ValidateRequest(
		validateNulls = true,
		validateEmpty = true,
		errorMessage = "결제 상태 변경 요청이 올바르지 않습니다.",
		exceptionType = SecurityException.class
	)
	@AuditLog(
		action = "PAYMENT_STATUS_CHANGE",
		resource = "PAYMENT",
		description = "🔒 중요: 결제 상태가 변경되었습니다",
		userIdExpression = "T(org.springframework.security.core.context.SecurityContextHolder).context.authentication?.principal?.userId ?: 'system'",
		resourceIdExpression = "#paymentId",
		includeRequestData = true,
		includeResponseData = true,
		level = AuditLog.Level.WARN
	)
	@Idempotent(
		keyExpression = "#paymentId + ':status_change:' + #request.status",
		keyPrefix = "payment_status",
		responseType = PaymentDetailResponse.class
	)
	@RetryOnFailure(
		maxAttempts = 3,
		backoffMillis = 500,
		exponentialBackoff = true,
		retryOn = {RuntimeException.class},
		noRetryOn = {SecurityException.class, IllegalArgumentException.class},
		logRetryAttempts = true
	)
	public ResponseEntity<BaseResponse<PaymentDetailResponse>> updatePaymentStatus(
			@Parameter(description = "결제 ID", required = true, example = "70000000-0000-0000-0000-000000000003")
			@PathVariable UUID paymentId,
			@io.swagger.v3.oas.annotations.parameters.RequestBody(
				description = "결제 상태 변경 요청",
				required = true,
				content = @Content(
					schema = @Schema(implementation = PaymentStatusUpdateRequest.class),
					examples = {
						@ExampleObject(
							name = "결제 승인",
							value = """
								{
								  "status": "PAID"
								}
								"""
						),
						@ExampleObject(
							name = "결제 실패",
							value = """
								{
								  "status": "FAILED"
								}
								"""
						),
						@ExampleObject(
							name = "결제 취소",
							value = """
								{
								  "status": "CANCELLED"
								}
								"""
						)
					}
				)
			)
			@Valid @RequestBody PaymentStatusUpdateRequest request) {
		com.popcorn.demo.domain.payment.service.PaymentCommandService.PaymentDetailResult result =
				paymentCommandService.updatePaymentStatus(paymentId, request.getStatus());
		return ok(toCommandDetailResponse(result));
	}

	@Operation(summary = "결제 삭제(소프트 삭제)", description = "결제 기록을 소프트 삭제합니다.")
	@ApiResponse(responseCode = "204", description = "결제 삭제 성공")
	@DeleteMapping("/payments/{paymentId}")
	@RateLimit(
		requests = 3,
		window = 3600,
		keyExpression = "T(org.springframework.web.context.request.RequestContextHolder).currentRequestAttributes().getRequest().getRemoteAddr()",
		errorMessage = "🚨 결제 삭제는 1시간에 3회만 허용됩니다.",
		algorithm = RateLimit.Algorithm.TOKEN_BUCKET
	)
	@ApiLogging(
		message = "🚨 결제 삭제",
		includeRequest = true,
		includeResponse = true,
		includeExecutionTime = true,
		level = ApiLogging.LogLevel.ERROR
	)
	@AuditLog(
		action = "PAYMENT_DELETE",
		resource = "PAYMENT",
		description = "🚨 위험: 결제 기록이 삭제되었습니다",
		userIdExpression = "T(org.springframework.security.core.context.SecurityContextHolder).context.authentication?.principal?.userId ?: 'anonymous'",
		resourceIdExpression = "#paymentId",
		includeRequestData = true,
		level = AuditLog.Level.ERROR
	)
	@ValidateRequest(
		validateNulls = true,
		errorMessage = "결제 삭제 요청이 올바르지 않습니다.",
		exceptionType = SecurityException.class
	)
	public ResponseEntity<Void> deletePayment(
			@Parameter(description = "결제 ID", required = true, example = "70000000-0000-0000-0000-000000000003")
			@PathVariable UUID paymentId) {
		paymentCommandService.deletePayment(paymentId);
		return noContent();
	}

	/**
	 * Query Service 결과를 응답 DTO로 변환 (QR 정보 포함)
	 */
	private PaymentDetailResponse toDetailResponse(PaymentDetailResult result) {
		return PaymentDetailResponse.builder()
				.paymentId(result.getPaymentId())
				.orderId(result.getOrderId())
				.method(result.getMethod() != null ? result.getMethod().name() : null)
				.status(result.getPaymentStatus() != null ? result.getPaymentStatus().name() : null)
				.amount(result.getAmount())
				.approvedAt(result.getApprovedAt())
				.createdAt(result.getCreatedAt())
				.updatedAt(result.getUpdatedAt())
				.orderStatus(result.getOrderStatus() != null ? result.getOrderStatus().name() : null)
				.qrAvailable(result.getQrAvailable())
				.qrCode(result.getQrCode())
				.qrExpiresAt(result.getQrExpiresAt())
				.build();
	}

	/**
	 * Command Service 결과를 응답 DTO로 변환 (QR 정보 없음)
	 */
	private PaymentDetailResponse toCommandDetailResponse(
			com.popcorn.demo.domain.payment.service.PaymentCommandService.PaymentDetailResult result) {
		return PaymentDetailResponse.builder()
				.paymentId(result.getPaymentId())
				.orderId(result.getOrderId())
				.method(result.getMethod() != null ? result.getMethod().name() : null)
				.status(result.getPaymentStatus() != null ? result.getPaymentStatus().name() : null)
				.amount(result.getAmount())
				.approvedAt(result.getApprovedAt())
				.createdAt(result.getCreatedAt())
				.updatedAt(result.getUpdatedAt())
				.orderStatus(result.getOrderStatus() != null ? result.getOrderStatus().name() : null)
				.qrAvailable(false) // Command 작업에서는 QR 정보 없음
				.qrCode(null)
				.qrExpiresAt(null)
				.build();
	}

	private PaymentListResponse.Item toListItem(PaymentDetailResult result) {
		return PaymentListResponse.Item.builder()
				.paymentId(result.getPaymentId())
				.orderId(result.getOrderId())
				.method(result.getMethod() != null ? result.getMethod().name() : null)
				.status(result.getPaymentStatus() != null ? result.getPaymentStatus().name() : null)
				.amount(result.getAmount())
				.approvedAt(result.getApprovedAt())
				.createdAt(result.getCreatedAt())
				.build();
	}
}

File: payment/src/main/kotlin/com/popcorn/payment/controller/PaymentController.kt
package com.popcorn.payment.controller

import com.popcorn.payment.dto.*
import com.popcorn.payment.exception.PaymentException
import com.popcorn.payment.service.TossPaymentCoroutineService
import com.popcorn.payment.service.PaymentCommandCoroutineService
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.Parameter
import io.swagger.v3.oas.annotations.media.Content
import io.swagger.v3.oas.annotations.media.Schema
import io.swagger.v3.oas.annotations.responses.ApiResponse as SwaggerApiResponse
import io.swagger.v3.oas.annotations.responses.ApiResponses
import io.swagger.v3.oas.annotations.tags.Tag
import jakarta.validation.Valid
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.util.*

/**
 * 결제 API 컨트롤러 (코루틴 기반)
 */
@Tag(name = "Payment API", description = "결제 관리 API")
@RestController
@RequestMapping("/api/v1/payments")
@CrossOrigin(origins = ["http://localhost:3000", "http://localhost:8080"])
class PaymentController(
    private val tossPaymentService: TossPaymentCoroutineService,
    private val paymentCommandService: PaymentCommandCoroutineService
) {

    private val log = LoggerFactory.getLogger(PaymentController::class.java)

    /**
     * 토스페이먼츠 결제 승인
     */
    @Operation(
        summary = "결제 승인",
        description = "토스페이먼츠를 통한 결제 승인을 처리합니다. 멱등성이 보장되어 중복 요청 시 동일한 결과를 반환합니다."
    )
    @ApiResponses(
        SwaggerApiResponse(responseCode = "200", description = "결제 승인 성공"),
        SwaggerApiResponse(responseCode = "400", description = "잘못된 요청 데이터"),
        SwaggerApiResponse(responseCode = "500", description = "서버 내부 오류")
    )
    @PostMapping("/confirm")
    suspend fun confirmPayment(
        @Valid @RequestBody
        @Parameter(description = "결제 승인 요청 정보", required = true)
        request: PaymentConfirmRequest
    ): ResponseEntity<ApiResponse<PaymentConfirmResponse>> = coroutineScope {

        log.info("💳 결제 승인 요청: paymentKey={}, orderId={}, amount={}원",
            request.paymentKey, request.orderId, request.amount)

        try {
            val result = tossPaymentService.confirmPayment(
                paymentKey = request.paymentKey,
                orderId = request.orderId,
                amount = request.amount
            )

            val response = PaymentConfirmResponse(
                paymentId = result.paymentId,
                paymentStatus = result.paymentStatus,
                orderStatus = result.orderStatus,
                orderId = result.orderId,
                orderNo = result.orderNo,
                amount = result.amount,
                approvedAt = result.approvedAt
            )

            log.info("✅ 결제 승인 성공: paymentId={}, amount={}원", result.paymentId, result.amount)
            ResponseEntity.ok(ApiResponse.success(response, "결제가 성공적으로 승인되었습니다."))

        } catch (e: PaymentException) {
            log.error("❌ 결제 승인 실패: paymentKey={}, error={}", request.paymentKey, e.message)
            ResponseEntity.status(HttpStatus.BAD_REQUEST)
                .body(ApiResponse.error(e.message ?: "결제 승인에 실패했습니다."))

        } catch (e: Exception) {
            log.error("❌ 결제 승인 중 예외 발생: paymentKey={}", request.paymentKey, e)
            ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(ApiResponse.error("시스템 오류가 발생했습니다. 잠시 후 다시 시도해주세요."))
        }
    }

    /**
     * 결제 취소
     */
    @PostMapping("/{paymentId}/cancel")
    suspend fun cancelPayment(
        @PathVariable paymentId: UUID,
        @Valid @RequestBody request: PaymentCancelRequest
    ): ResponseEntity<ApiResponse<PaymentCancelResponse>> {

        log.info("🔄 결제 취소 요청: paymentId={}, reason={}", paymentId, request.cancelReason)

        try {
            // 결제 정보로부터 주문 ID 조회
            val paymentDetail = paymentCommandService.getLatestPaymentByOrderId(paymentId) // 임시로 paymentId 사용

            val result = tossPaymentService.cancelPayment(
                orderId = paymentDetail.orderId ?: paymentId, // 임시 처리
                cancelReason = request.cancelReason
            )

            val response = PaymentCancelResponse(
                paymentId = result.paymentId,
                orderId = result.orderId,
                cancelAmount = result.cancelAmount,
                status = result.status,
                cancelReason = result.cancelReason
            )

            log.info("✅ 결제 취소 성공: paymentId={}, cancelAmount={}원", result.paymentId, result.cancelAmount)
            return ResponseEntity.ok(ApiResponse.success(response, "결제가 성공적으로 취소되었습니다."))

        } catch (e: PaymentException) {
            log.error("❌ 결제 취소 실패: paymentId={}, error={}", paymentId, e.message)
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                .body(ApiResponse.error(e.message ?: "결제 취소에 실패했습니다."))

        } catch (e: Exception) {
            log.error("❌ 결제 취소 중 예외 발생: paymentId={}", paymentId, e)
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(ApiResponse.error("시스템 오류가 발생했습니다. 잠시 후 다시 시도해주세요."))
        }
    }

    /**
     * 결제 생성
     */
    @PostMapping
    suspend fun createPayment(
        @Valid @RequestBody request: PaymentCreateRequest
    ): ResponseEntity<ApiResponse<PaymentCreateResponse>> {

        log.info("📝 결제 생성 요청: orderId={}, method={}, amount={}원",
            request.orderId, request.paymentMethod, request.amount)

        try {
            val result = paymentCommandService.createPayment(
                orderId = request.orderId,
                paymentMethod = request.paymentMethod,
                amount = request.amount,
                rawPayload = ""
            )

            val response = PaymentCreateResponse(
                paymentId = result.paymentId,
                orderId = request.orderId,
                amount = result.amount,
                status = result.status,
                paymentMethod = request.paymentMethod,
                createdAt = result.createdAt
            )

            log.info("✅ 결제 생성 성공: paymentId={}", result.paymentId)
            return ResponseEntity.status(HttpStatus.CREATED)
                .body(ApiResponse.success(response, "결제가 성공적으로 생성되었습니다."))

        } catch (e: PaymentException) {
            log.error("❌ 결제 생성 실패: orderId={}, error={}", request.orderId, e.message)
            return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                .body(ApiResponse.error(e.message ?: "결제 생성에 실패했습니다."))

        } catch (e: Exception) {
            log.error("❌ 결제 생성 중 예외 발생: orderId={}", request.orderId, e)
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(ApiResponse.error("시스템 오류가 발생했습니다. 잠시 후 다시 시도해주세요."))
        }
    }

    /**
     * 결제 상세 조회
     */
    @GetMapping("/{paymentId}")
    suspend fun getPayment(
        @PathVariable paymentId: UUID
    ): ResponseEntity<ApiResponse<PaymentDetailResponse>> {

        log.debug("🔍 결제 조회 요청: paymentId={}", paymentId)

        try {
            // 임시로 구현 (실제로는 별도 조회 서비스 필요)
            val result = paymentCommandService.getLatestPaymentByOrderId(paymentId)

            val response = PaymentDetailResponse(
                paymentId = result.paymentId,
                orderId = result.orderId ?: UUID.randomUUID(),
                amount = result.amount,
                status = result.status,
                paymentMethod = "CARD", // 임시값
                createdAt = java.time.LocalDateTime.now(),
                approvedAt = result.approvedAt,
                updatedAt = java.time.LocalDateTime.now()
            )

            return ResponseEntity.ok(ApiResponse.success(response))

        } catch (e: PaymentException) {
            log.error("❌ 결제 조회 실패: paymentId={}, error={}", paymentId, e.message)
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(ApiResponse.error(e.message ?: "결제 정보를 찾을 수 없습니다."))

        } catch (e: Exception) {
            log.error("❌ 결제 조회 중 예외 발생: paymentId={}", paymentId, e)
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(ApiResponse.error("시스템 오류가 발생했습니다. 잠시 후 다시 시도해주세요."))
        }
    }

    /**
     * 주문의 결제 목록 조회
     */
    @GetMapping("/orders/{orderId}")
    suspend fun getPaymentsByOrderId(
        @PathVariable orderId: UUID
    ): ResponseEntity<ApiResponse<PaymentListResponse>> = coroutineScope {

        log.debug("🔍 주문 결제 목록 조회: orderId={}", orderId)

        try {
            // 비동기로 결제 목록과 통계 조회
            val paymentsDeferred = async {
                // 임시 구현 - 실제로는 PaymentRepository의 조회 메서드 사용
                listOf<PaymentDetailResponse>()
            }

            val statsDeferred = async {
                // 결제 통계 계산
                Pair(0, 0L) // (건수, 총금액)
            }

            val payments = paymentsDeferred.await()
            val (totalCount, totalAmount) = statsDeferred.await()

            val response = PaymentListResponse(
                payments = payments,
                totalCount = totalCount,
                totalAmount = totalAmount
            )

            ResponseEntity.ok(ApiResponse.success(response))

        } catch (e: Exception) {
            log.error("❌ 주문 결제 목록 조회 실패: orderId={}", orderId, e)
            ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                .body(ApiResponse.error("시스템 오류가 발생했습니다."))
        }
    }

    /**
     * 헬스 체크
     */
    @GetMapping("/health")
    suspend fun healthCheck(): ResponseEntity<Map<String, Any>> {
        val healthInfo = mapOf(
            "status" to "UP",
            "timestamp" to java.time.LocalDateTime.now(),
            "service" to "payment-service"
        )
        return ResponseEntity.ok(healthInfo)
    }
}